components/tidb_query_executors/src/index_lookup_executor.rs

```diff
@@ -0,0 +1,2287 @@
+        self.phase = IndexLookUpPhase::TableLookUp(TableLookUpState {
+            table_task_iter: Some(builder.build_iterator(&mut ctx, results)?),
+            table_scan: None,
+        });
+
+        if ctx.warnings.warning_cnt > 0 {
+            warnings.merge(&mut ctx.warnings);
+        }
+
+        Ok(())
+    }
+
+    #[inline]
+    fn finish_table_task_iter(&mut self) -> Result<Vec<BatchExecuteResult>> {
+        let table_lookup = self.phase.mut_table_lookup_or_err()?;
+        if let Some(iter) = table_lookup.table_task_iter.take() {
+            let mut results = iter.take_left_results();
+            self.intermediate_results.append(&mut results);
+            Ok(results)
+        } else {
+            Ok(vec![])
+        }
+    }
+
+    #[inline]
+    fn step_to_index_scan(&mut self) -> Result<()> {
+        let results = self.finish_table_task_iter()?;
+        let mut index_scan = IndexScanState::default();
+        // reuses results
+        index_scan.results = results;
+        self.phase = IndexLookUpPhase::IndexScan(index_scan);
+        Ok(())
+    }
+
+    #[inline]
+    fn step_to_done_from_table_lookup(&mut self) -> Result<()> {
+        self.finish_table_task_iter()?;
+        self.phase = IndexLookUpPhase::Done;
+        Ok(())
+    }
+
+    async fn on_phase_index_scan(&mut self, scan_rows: usize) -> Result<BatchExecuteResult> {
+        if self.src_is_drained.stop() {
+            return Err(other_err!(
+                "src is already drained, cannot continue index scan"
+            ));
+        }
+
+        let state = self.phase.mut_index_scan_or_err()?;
+        let mut result = self.src.next_batch(scan_rows).await;
+        let src_drained = match result.is_drained {
+            Ok(is_drained) if is_drained.stop() => {
+                self.src_is_drained = is_drained;
+                true
+            }
+            _ => false,
+        };
+
+        let mut output_result = Self::empty_result(BatchExecIsDrain::Remain);
+        mem::swap(&mut result.is_drained, &mut output_result.is_drained);
+        if result.warnings.warning_cnt > 0 {
+            mem::swap(&mut result.warnings, &mut output_result.warnings);
+        }
+        if self.force_no_index_lookup {
+            if !result.logical_rows.is_empty() {
+                self.intermediate_results.push(result);
+            }
+            if src_drained {
+                self.phase = IndexLookUpPhase::Done;
+            }
+            return Ok(output_result);
+        }
+
+        if !result.logical_rows.is_empty() {
+            state.row_count += result.logical_rows.len();
+            state.results.push(result);
+        }
+
+        if src_drained && state.results.is_empty() {
+            self.phase = IndexLookUpPhase::Done;
+            return Ok(output_result);
+        }
+
+        if src_drained || state.row_count >= self.table_lookup_batch_size {
+            self.step_to_table_lookup(&mut output_result.warnings)?;
```

Comment:
If the `src_drained` is true, the state machine should exit after table lookup, but the `src_drained` is not saved, is it a little different from the state machine figure, seems the red circle is unreachable in the implementation.

<img width="813" height="405" alt="image" src="https://github.com/user-attachments/assets/523614da-9189-4287-8480-ce2d4dcb8d19" />

Reply:
The state is saved here:

https://github.com/tikv/tikv/pull/18869/files#diff-a1b1710ceff528e5d4841d11516456e14d0e2a028f088f3c673ce6798e9d9028R241

```
        let src_drained = match result.is_drained {
            Ok(is_drained) if is_drained.stop() => {
                self.src_is_drained = is_drained;
                true
            }
            _ => false,
        };
```